src/algebra/char_p/localization.lean

```diff
@@ -0,0 +1,68 @@
+/-
+Copyright (c) 2021 Jon Eugster. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jon Eugster
+-/
+import algebra.char_p.basic
+import ring_theory.localization
+
+
+/-!
+# Characteristics of field of fractions
+
+In this file we describe the characteristics of the field of fractions of an integral domain.
+
+
+## Main results
+
+If `R` is an `integral_domain` then the `fraction_ring R` has the same characteristic as `R`.
+
+
+## TODO
+
+* Could generalise this to any `R`-algebra over a `comm_semiring R`
+  with injective algebra map.
+
+-/
+
+namespace fraction_ring
+
+variables (R: Type*) [integral_domain R] (p: ℕ)
+
+/--
+  If `R` has characteristic `p`, then so does `fraction_ring R`.
+-/
+instance char_p [h: char_p R p]: char_p (fraction_ring R) p :=
+{ cast_eq_zero_iff :=
+  begin
+    intro x,
+    have hR:= char_p.cast_eq_zero_iff R p x,
+    split,
+    {
+      intro hp,
+      apply hR.elim_left,
+      change (coe : ℕ → R) x = 0,
+      have q := (fraction_map.injective (fraction_ring.of R)),
+      unfold function.injective at q,
+      apply q,
+      simp,
+      exact hp,
+    },
+    {
+      intro hp,
+      have hR' := hR.elim_right hp,
+      change (algebra_map ℕ (fraction_ring R)) x = (algebra_map ℕ (fraction_ring R)) 0,
+      rw is_scalar_tower.algebra_map_apply ℕ R (fraction_ring R) x,
+      change (algebra_map ℕ R) x = (algebra_map ℕ R) 0 at hR',
+      rw hR',
+      simp,
+    }
+  end}
+
+/--
+  If `R` has characteristic `0`, then so does `fraction_ring R`.
+-/
+instance char_zero [h: char_zero R]: char_zero (fraction_ring R) :=
+char_p.char_p_to_char_zero (fraction_ring R)
```

Comment:
It's better not to name the typeclass hypothesis if you're not going to use the name:
```suggestion
/-- If `R` has characteristic `0`, then so does `fraction_ring R`. -/
instance [char_zero R] : char_zero (fraction_ring R) :=
char_p.char_p_to_char_zero (fraction_ring R)
```
Lean will generate the instance name for you too.